Add Unix version of coreconsole.

adityamandaleeka opened the pull request:
Created a Unix version of coreconsole which, like the Windows version,
is similar to corerun but runs the managed binary with the same name as
the executable (but with the DLL extension). Moved code that is common
between corerun and coreconsole to coreruncommon.

Fix #747

adityamandaleeka commented:
Looks like the Mac build failed. I just need to include the same workaround that's used in unixcorerun in order to get past this XCode SDK issue with __debug.

adityamandaleeka commented:
/cc @janvorli @sergiy-k @jkotas 
Please review this when you have some time.

kangaroo commented:
I don't like the fact that we now have 2 of the exact same hack-around for the Xcode 6.3 problem.  Could you please move this to an appropriate place where we can share it with the other one?  I'd really not like to end up with 10 different copies of this hack to track down once apple fixes their headers.

janvorli commented:
LGTM modulo @kangaroo's comment.

akoeplinger commented on src/coreclr/hosts/unixcoreconsole/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+
+target_link_libraries(coreconsole 
+    dl
+)
```
This needs the same check for freebsd that was just merged in https://github.com/dotnet/coreclr/pull/791

adityamandaleeka commented:
Thanks everyone.

I agree with all of the comments. About the XCode 6.3 issue, it appears that Apple has fixed their headers in the 6.3.1 update. If we update our build machines to 6.3.1, we should be able to remove the workaround entirely.

I'll update the PR soon.

sergiy-k commented:
Thank you, @adityamandaleeka! I agree that updating the build machines to 6.3.1 and removing the workaround would be the best solution.

@mmitche, what do you think?

mmitche commented:
Works for me, I'll get on it this afternoon.

adityamandaleeka commented:
Thanks @mmitche. 

#861 removed the workaround for the XCode 6.3 problem. I've updated this PR accordingly (and also updated unixcoreruncommon to reflect changes made to unixcorerun since last week), and will merge it after a successful build unless there are further comments.
